[SPARK-25809][K8S][TEST] New K8S integration testing backends

rvesse opened the pull request:
## What changes were proposed in this pull request?

Currently K8S integration tests are hardcoded to use a `minikube` based backend.  `minikube` is VM based so can be resource hungry and also doesn't cope well with certain networking setups (for example using Cisco AnyConnect software VPN `minikube` is unusable as it detects its own IP incorrectly).

This PR Adds a new K8S integration testing backend that allows for using the Kubernetes support in [Docker for Desktop](https://blog.docker.com/2018/07/kubernetes-is-now-available-in-docker-desktop-stable-channel/).  It also generalises the framework to be able to run the integration tests against an arbitrary Kubernetes cluster.

To Do:

- [x] General Kubernetes cluster backend
- [x] Documentation on Kubernetes integration testing
- [x] Testing of general K8S backend
- [x] Check whether change from timestamps being `Time` to `String` in Fabric 8 upgrade needs additional fix up

## How was this patch tested?

Ran integration tests with Docker for Desktop and all passed:

![screen shot 2018-10-23 at 14 19 56](https://user-images.githubusercontent.com/2104864/47363460-c5816a00-d6ce-11e8-9c15-56b34698e797.png)

Suggested Reviewers: @ifilonenko @srowen

rvesse committed:
[SPARK-25809][K8S] Docker for desktop K8S integration backend

Adds a new K8S integration testing backend that allows for using the
Kubernetes support in Docker for Desktop

AmplabJenkins commented:
Can one of the admins verify this patch?

rvesse commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala:
```diff
@@ -0,0 +1,68 @@
+    // TODO: If we were using Fabric 8 client 3.1.0 then we could
+    // instead just use the overload of autoConfigure() that takes the
+    // desired context avoiding the need to interact with kubectl at all
+    initialContext = getCurrentContext
```
Would people be opposed to bumping the K8S client to 3.1.0 (or even a more recent 3.1.x release) as this would simplify the code significantly and avoid any need to shell out to `kubectl`?

liyinan926 replied:
No, this is a welcomed. If it's not a significant change to upgrade to 3.1.0, we probably should make it part of this PR.

srowen commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala:
```diff
@@ -0,0 +1,68 @@
+package org.apache.spark.deploy.k8s.integrationtest.backend.docker
```
(needs an Apache copyright header)

rvesse replied:
Fixed

srowen commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala:
```diff
@@ -30,14 +30,16 @@ private[spark] trait IntegrationTestBackend {
 private[spark] object IntegrationTestBackendFactory {
   val deployModeConfigKey = "spark.kubernetes.test.deployMode"
 
+  val backendMinikube = "minikube"
```
Nit: should these be private?

srowen commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  override def cleanUp(): Unit = {
+    super.cleanUp()
```
Not sure, but do you want to do the superclass cleanup afterwards, after you finish cleanup in this class?

srowen commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala:
```diff
@@ -0,0 +1,68 @@
+
+  private def executeKubectl(args: String*): Seq[String] = {
+    ProcessUtils.executeProcess(
+      Array("kubectl") ++ args, KUBECTL_STARTUP_TIMEOUT_SECONDS, true)
```
Total nit, but you can write `"kubectl" +: args`

srowen commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala:
```diff
@@ -0,0 +1,68 @@
+import org.apache.spark.deploy.k8s.integrationtest.ProcessUtils
+import org.apache.spark.deploy.k8s.integrationtest.backend.IntegrationTestBackend
+
+private[spark] object DockerForDesktopBackend extends IntegrationTestBackend {
```
Is it OK that this is basically a singleton? it has state. I'm just wondering if that could cause a problem if used across many tests.

rvesse replied:
With the latest commits the only state maintained is the K8S client which is reusable and there is no longer any need to modify the on-disk state via `kubectl` by adopting a newer Fabric 8 client library version

srowen reviewed (commented):
I don't have substantive comments, just minor ones. Could this be why the k8s tests failed on a fresh Ubuntu install? I didn't do any k8s setup (should I?)

rvesse commented:
@srowen Yes there are a lot of assumptions made by the integration tests that are not documented anywhere and I figured out by digging in the code and POMs.

Broadly speaking right now to run integration tests you need the following:

- A runnable distribution of Spark built
- `minikube` installed, configured with at least 4 CPUs and 8 GB memory and started

Probably the next thing I will do tomorrow on this is to improve the documentation.  There is a `README.md` in the integration testing module but it omits various details I have encountered working with these recently.

skonto commented:
@rvesse this capability was in the old integration tests repo [here](https://github.com/apache-spark-on-k8s/spark-integration/blob/master/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/CloudTestBackend.scala).

In our[ fork](https://github.com/lightbend/spark/tree/spark-k8s-2.4-snapshot/resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend) we have merged  that part back, can we add that too? I find it really useful to test real platforms like Openshift etc.

skonto commented:
@rvesse also the following memory limit works for me:

`minikube config set memory 4096` so maybe g its too much.

rvesse commented:
@skonto Yep, I plan to do that tomorrow

At least for my `minikube` instance I found 4g insufficient and a couple of tests would fail because their pods didn't get scheduled.  8g is probably over allocating, somewhere in-between likely works also

skonto commented:
@rvesse I use no vm:
`sudo -E ./minikube start --vm-driver=none`

ifilonenko commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala:
```diff
@@ -30,14 +30,16 @@ private[spark] trait IntegrationTestBackend {
+    deployMode match {
+      case `backendMinikube` => MinikubeTestBackend
+      case `backendDockerForDesktop` => DockerForDesktopBackend
+      case _ => throw new IllegalArgumentException("Invalid " + deployModeConfigKey + ": " + deployMode)
```
NIT: line too long, split

ifilonenko commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala:
```diff
@@ -30,14 +30,16 @@ private[spark] trait IntegrationTestBackend {
+      .getOrElse(backendMinikube)
+    deployMode match {
+      case `backendMinikube` => MinikubeTestBackend
+      case `backendDockerForDesktop` => DockerForDesktopBackend
```
Can you add a README to explain how to run with docker-for-desktop, by specifying --deploy-mode

ifilonenko reviewed (commented):
Thank you so much for this change. This looks pretty good for me, pending small nits and README. 
I tried this change out and it works like a charm. I will have @mccheah and @liyinan926 look at this for merging upon their LGTM. 

Can we make this PR only target docker-for-desktop as I personally would like this to go in. Could a generalization come in a followup? Would like to hear the opinion of the committers

liyinan926 commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/ProcessUtils.scala:
```diff
@@ -40,7 +40,8 @@ object ProcessUtils extends Logging {
       s"Timed out while executing ${fullCommand.mkString(" ")}")
-    assert(proc.exitValue == 0, s"Failed to execute ${fullCommand.mkString(" ")}")
+    assert(proc.exitValue == 0,
+      s"Failed to execute ${fullCommand.mkString(" ")}${if (dumpErrors) outputLines.mkString("\n")}")
```
Should add a `\n` before the first line of the error output.

rvesse committed:
[SPARK-25809][K8S] Integration test improvements

- Adds cloud and cloud-url backends
- Improves logic in integration test scripts to cover new properties
- Move constants into TestConstants
- Use Config.autoConfigure() with context for simpler implementation
  without the need to shell out
- Improve ReadMe with more detail on test backends, configuration and
  assumptions

rvesse committed:
[SPARK-25809][K8S] Add and use constants for config

Clarify code by using appropriate constants from TestConstants rather
than manually reusing strings

rvesse commented:
@ifilonenko I have done the generalisation today since it was fairly trivial and it actually resolves a number of concerns about the first pass implementation

@skonto I have restored the cloud backend as requested, I ended up having two separate cloud backends.  One using the original code that simply takes a master URL and another that takes a desired K8S context and auto-configures itself from the K8S config file and the context specified.

rvesse commented:
Ran successfully against one of our dev K8S clusters today:

![screen shot 2018-10-25 at 17 39 40](https://user-images.githubusercontent.com/2104864/47516269-fd2e1480-d87c-11e8-80f9-2325edffe066.png)

Will continue onto testing against one of our larger production clusters

liyinan926 commented on resource-managers/kubernetes/integration-tests/scripts/setup-integration-test-env.sh:
```diff
@@ -71,19 +71,36 @@ if [[ $IMAGE_TAG == "N/A" ]];
+      fi
+      ;;
+
+     docker-for-desktop)
```
The indention from this point seems to be off.

rvesse replied:
Now corrected

liyinan926 commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:
```diff
@@ -27,4 +27,36 @@ object Utils extends Logging {
     try f.apply(resource) finally resource.close()
   }
+
+  def checkAndGetK8sMasterUrl(rawMasterURL: String): String = {
```
Is this the same as https://github.com/apache/spark/blob/a2f502cf53b6b00af7cb80b6f38e64cf46367595/core/src/main/scala/org/apache/spark/util/Utils.scala#L2725?

rvesse replied:
Yes, will fix in next commit

rvesse replied:
Now fixed

liyinan926 commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala:
```diff
@@ -28,16 +30,16 @@ private[spark] trait IntegrationTestBackend {
+      .getOrElse(BACKEND_MINIKUBE)
+    deployMode match {
+      case BACKEND_MINIKUBE => MinikubeTestBackend
+      case BACKEND_CLOUD => new KubeConfigBackend(null)
```
Can we combine `KubeConfigBackend` and `CloudTestBackend` into one? The only difference is how the kubeconfig gets constructed. I guess you can decide which way to go based on the presence of the `CONFIG_KEY_KUBE_MASTER_URL` property.

rvesse replied:
Yes, done this in the latest commit

mccheah commented on resource-managers/kubernetes/integration-tests/README.md:
```diff
@@ -13,15 +13,45 @@ The simplest way to run the integration tests is to install and run Minikube, th
+## Using a different backend
+
+The integration test backend i.e. the K8S cluster used for testing is controlled by the `--deploy-mode` option.  By default this
+is set to `minikube`, the available backends are their perquisites are as follows.
```
`By default this is set to minikube. The available backends and their prerequisites are described below.`

rvesse replied:
Fixed

mccheah commented on resource-managers/kubernetes/integration-tests/README.md:
```diff
@@ -41,12 +71,127 @@ The Spark code to test is handed to the integration test system via a tarball. H
+# Running the Test Directly
+
+If you prefer to run just the integration tests directly then you can customise the behaviour via properties passed to Maven using the
+`-Dproperty=value` option e.g.
```
`If you prefer to run just the integration tests directly, then you can customize the behaviour via passing system properties to Maven. For example:`

rvesse replied:
Fixed

liyinan926 commented:
@rvesse I want to take back my comment on bumping the version of the fabric8 client in the same PR. Based on the discussion so far, it seems more appropriate to do that in a separate PR, like in #22820. Can you revert the changes to bump the version? Sorry about that.

mccheah commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala:
```diff
@@ -0,0 +1,58 @@
+import org.apache.spark.deploy.k8s.integrationtest.backend.IntegrationTestBackend
+import org.apache.spark.internal.Logging
+
+private[spark] class KubeConfigBackend(var context: String)
```
Just use a default value for `context` instead of checking for null below.

rvesse replied:
Refactored to avoid this

mccheah commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala:
```diff
@@ -0,0 +1,58 @@
+
+  override def initialize(): Unit = {
+    // Auto-configure K8S client from K8S config file
+    System.setProperty(Config.KUBERNETES_AUTH_TRYKUBECONFIG_SYSTEM_PROPERTY, "true");
```
Setting system properties is a little worrisome, for example what if something else sets this system property as well? Would much prefer to treat system properties as immutable after the JVM starts.

rvesse replied:
Looked over what Fabric 8 client was doing behind the scenes and realised the values I was setting explicitly are the defaults anyway so removed this unnecessary manipulation

rvesse committed:
[SPARK-25809][K8S] Add TODOs for open issue

Identified an open issue with spark-submit only ever auto-configuring
itself from the current KUBECONFIG context which results in errors if
trying to use a non-default context for integration tests.

rvesse commented:
@liyinan926 I will rebase and squash appropriately once PR #22820 is merged

rvesse committed:
[SPARK-25809][K8S] Further simplify code

- Combine cloud backends into one
- Remove unecessary manipulation of System properties
- Documentation tweaks
- Ensure we clean up Spark Master URL when overriding master URL in
  KubeConfigBackend
- Reuse utils method from Spark Core rather than duplicating in
  integration tests module
- Fix Bash indentation issues
- Insert extra new line prior to error output

rvesse committed:
[SPARK-25809][K8S] Add extra profile to README

liyinan926 commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala:
```diff
@@ -0,0 +1,70 @@
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils.checkAndGetK8sMasterUrl
+
+private[spark] class KubeConfigBackend(var context: String)
```
What if the input `context` is null, e.g., if `CONFIG_KEY_KUBE_CONFIG_CONTEXT` is not defined?

liyinan926 commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala:
```diff
@@ -28,16 +30,15 @@ private[spark] trait IntegrationTestBackend {
+      .getOrElse(BACKEND_MINIKUBE)
+    deployMode match {
+      case BACKEND_MINIKUBE => MinikubeTestBackend
+      case BACKEND_CLOUD => new KubeConfigBackend(System.getProperty(CONFIG_KEY_KUBE_CONFIG_CONTEXT))
```
What if `CONFIG_KEY_KUBE_CONFIG_CONTEXT` is not defined?

rvesse replied:
The Fabric 8 client treats `null` as meaning use the users current context from their `KUBECONFIG` file - https://github.com/fabric8io/kubernetes-client/blob/v4.1.0/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java#L467

liyinan926 replied:
Got it.

liyinan926 reviewed (commented):
LGTM. Let us know when you finish your testing.

rvesse commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala:
```diff
@@ -42,6 +42,9 @@ private[spark] object SparkKubernetesClientFactory {
       defaultServiceAccountCaCert: Option[File]): KubernetesClient = {
+
+    // TODO Support configurable context
+
```
Note to reviewers.  In testing this I noticed one issue is that when you have multiple contexts defined in your `KUBECONFIG` file that while this PR adds the ability for the integration test machinery to use an arbitrary context the submission client will always use your current context.  Therefore if your current context does not match the requested context the integration tests will fail/return false positives because the client and the test harness will use different contexts.

I have filed SPARK-25887 for addressing this since the client change seemed like it should be a separate issue and PR.  I will provide a PR for that tomorrow.

rvesse commented:
Did a bunch more testing on our internal K8S clusters today after rebasing this onto master. I am now happy that this is ready for final review and merging so I have removed the `[WIP]` tag

ifilonenko commented:
test this please

SparkQA commented:
**[Test build #98277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98277/testReport)** for PR 22805 at commit [`41af220`](https://github.com/apache/spark/commit/41af2204db8f85f4ebdad4538a26f24e269a45ee).

liyinan926 reviewed (approved):
LGTM. Thanks!

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/4636/

ifilonenko reviewed (approved):
LGTM (pending successful tests), thank you for this!

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/4636/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4636/
Test FAILed.

ifilonenko commented:
retest this please

SparkQA commented:
**[Test build #98279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98279/testReport)** for PR 22805 at commit [`41af220`](https://github.com/apache/spark/commit/41af2204db8f85f4ebdad4538a26f24e269a45ee).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/4638/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4638/
Test PASSed.

liyinan926 commented:
@mccheah since he had some comments.

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala:
```diff
@@ -63,6 +66,8 @@ private[spark] object SparkKubernetesClientFactory {
     val dispatcher = new Dispatcher(
       ThreadUtils.newDaemonCachedThreadPool("kubernetes-dispatcher"))
+
+    // TODO Create builder in a way that respects configurable context
```
Can we add a Spark ticket here?

rvesse replied:
SPARK-25887, I'll update the comment to reference it explicitly

mccheah reviewed (commented):
Just want to make sure we're tracking one follow-up item explicitly, once we add the comment change we can merge. Don't need to specifically wait for CI before the next merge either since it should just be a code comment change.

SparkQA commented:
**[Test build #98277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98277/testReport)** for PR 22805 at commit [`41af220`](https://github.com/apache/spark/commit/41af2204db8f85f4ebdad4538a26f24e269a45ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98277/
Test PASSed.

SparkQA commented:
**[Test build #98279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98279/testReport)** for PR 22805 at commit [`41af220`](https://github.com/apache/spark/commit/41af2204db8f85f4ebdad4538a26f24e269a45ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98279/
Test PASSed.

rvesse committed:
[SPARK-25809][K8S] Reference follow up ticket in comments

rvesse commented:
@mccheah I have updated the comment to reference the follow up issue and opened a PR for that as #22904.  Can we go ahead and merge now?

mccheah commented:
Yes, this can be merged. Going into master now.
